src/main.rs

```diff
@@ -6,13 +11,19 @@ use std::{
 
 use serde::{Deserialize, Serialize};
 use spectre::{graph::AGraph, graph::Graph};
+use ziggurat_core_geoip::geoip::GeoIPInfo;
+use ziggurat_core_geoip::providers::ip2loc::Ip2LocationService;
+use ziggurat_core_geoip::providers::ipgeoloc::{BackendProvider, IpGeolocateService};
 
+//TODO(asmie): there is some redundancy here as we have ip in the Node structure and one more time
+// in the GeoIPInfo structure.
 #[derive(Default, Clone, Serialize, Deserialize)]
 pub struct Node {
     ip: String,
```

Comment:
Why `String` instead of `IpAddr`?

Reply:
We can ask @kylegranger as I've not changed this.
Inside the GeoIPInfo (ziggurat-core) structure ip field is IpAddr type:
https://github.com/runziggurat/ziggurat-core/blob/db38b592ee67de60207e7747d4a09e9aa4ece611/ziggurat-core-geoip/src/geoip.rs#L17

Reply:
I kept the String type from NetworkSummary
https://github.com/runziggurat/ziggurat-core/blob/main/ziggurat-core-crawler/src/summary.rs#L24
At the time, I made it a string when we went from `SocketAddr` (with port number) to just plain ip address (was not aware of `IpAddr` type)
If we use `IpAddr`, it should probably be changed there, too.

Reply:
regarding the redundancy, I had an idea.
```
pub struct GeoInfo {
    /// Country name (long name)
    pub country: Option<String>,
    /// City name
    pub city: Option<String>,
    /// Latitude
    pub latitude: Option<f64>,
    /// Longitude
    pub longitude: Option<f64>,
    /// Timezone of the IP
    pub timezone: Option<String>,
    /// ISP name (unavailable for some providers)
    pub isp: Option<String>,
}

pub struct GeoIPInfo {
    /// IP address
    pub ip: IpAddr,
    /// GeoInfo struct
    pub geo_info: GeoInfo
}
```
And in the state.json, only the GeoInfo gets added to the node.
```
pub struct Node {
    ip: String,
    betweenness: f64,
    closeness: f64,
    num_connections: usize,
    geolocation: Option<GeoInfo>,
}
```

Reply:
Change done in ziggurat-core.